### ISSUE TITLE
Is there a way to retrain face recognition (dlib/compreface) without restarting Viseron?

### ISSUE THREAD

**Issue by tantonj** [roflcoopter/viseron#839]
Hello,

I'm using Viseron with dlib and love it so far! Forgive me if I am putting this in the wrong spot as I don't really see it as an "Issue". I just have a question how to do something and this seemed like the best place to do so.

I would like my own software to be able to add a new person to the face recognition path and have viseron retrain the face recognition model without restarting viseron. I'm wondering if this is possible.

I know I can change the face recognition path to anywhere I want so I can easily write any files I want in there. (Though my preference would be to do this remotely through a viseron endpoint so that Viseron can sit on it's own server without any middleware on it). But once doing so, Viseron appears to need to restart in order to retrain the model and be able to detect the new face.

I've also experimented with Compreface, (I still need to do more, but stopped because dlib was working). I was having trouble with it detecting a face. From what I understand it goes off of the faces in the same Viseron path and NOT the faces added to any application created inside Compreface itself. Compreface has the ability to retrain when a new person is added (if I am not mistaken). So is there a way to use the faces inside Compreface instead of the file structure in Viseron?

I appreciate any response even if it is that Viseron can and will not do any of these things. Worst case scenario I can create a small api server to receive an image and write it appropriately and use the websocket to restart Viseron when a new person is added. This of coarse means viseron will be effectively going down momentarily every time a new person is added to the system.

Thanks,

P.S. another option, if you point me in the right direction I can experiment with building Viseron from code and attempt a PR where an endpoint will do exactly this (receive face/name, update files, retrain model).

**Comment by roflcoopter** [comment]
> I would like my own software to be able to add a new person to the face recognition path and have viseron retrain the face recognition model without restarting viseron. I'm wondering if this is possible.

Not possible sadly, but there is no technical limitation or anything like that. I simply have not had the time to do it :(



> I've also experimented with Compreface, (I still need to do more, but stopped because dlib was working). I was having trouble with it detecting a face. From what I understand it goes off of the faces in the same Viseron path and NOT the faces added to any application created inside Compreface itself. Compreface has the ability to retrain when a new person is added (if I am not mistaken). So is there a way to use the faces inside Compreface instead of the file structure in Viseron?

As long as the persons trained in Compreface is the same as the foldernames in the Viseron config i think it should work.
So if you train a face in Compreface called `john`, and create an empty dir called `/config/face_recognition/faces/john/` and set `train: false` for `compreface` it should work.

Can you try that out?



> P.S. another option, if you point me in the right direction I can experiment with building Viseron from code and attempt a PR where an endpoint will do exactly this (receive face/name, update files, retrain model).

I would love that! I am always looking for more contributions.

**Comment by tantonj** [comment]
Thanks for the reply,

One last question from what you responded with.

> So if you train a face in Compreface called john, and create an empty dir called /config/face_recognition/faces/john/ and set train: false for compreface it should work.

If Viseron is running and there are 3 foldernames in the Viseron config and train is false, we add a new person/image to compreface, and create a new folder in the config to match that person, will Viseron need to restart in order to match that person? (I assume yes based on how binary facial recognition events are populated).

Thanks,

**Comment by tantonj** [comment]
I've managed to implement all the features I needed and tested them successfully. 
I haven't finished fixing all the linting errors, so I haven't been able to commit yet to my fork. Once I get through them I'll submit a pr.
I've implemented a new compreface config variable called "use_subjects" which defaults to false.
When set to true it blocks the binary face recognition entities from being created based on the folder structure. Instead when compreface initializes it pulls all subjects from compreface and generates the entities with them.
I then created a new endpoint at "api/v1/compreface/update_subjects", this will fetch the subjects from compreface and add any face recognition entities that are missing. 
This way a user can manage subjects through compreface's API and/or SDK, then call the endpoint when updated and be able to match new subjects without restarting viseron

This all seems to work well. However, I may need to rework how both the API and the compreface component do this. Linting is complaining about me accessing protected members (from self._vis). Which I'm definitely doing. If you are willing to email me at tantonj@protonmail.com I'd like to ask some advice on how best to refactor this so I can do it appropriately.
